iree/turbine/kernel/wave/runtime/runtime.cpp

```diff
@@ -0,0 +1,118 @@
+#include <nanobind/nanobind.h>
+#include <nanobind/stl/bind_vector.h>
+#include <nanobind/stl/string.h>
+#include <torch/torch.h>
+#include <torch/extension.h>
+#include <hip/hip_runtime.h>
+#include <hip/hip_runtime_api.h>
+#include <ATen/hip/HIPContext.h>
+#include <unordered_map>
+
+namespace nb = nanobind;
+
+#define HIP_CHECK_EXC(expr)                                                                       \
+    do                                                                                            \
+    {                                                                                             \
+        hipError_t e = (expr);                                                                    \
+        if (e)                                                                                    \
+        {                                                                                         \
+            const char *errName = hipGetErrorName(e);                                             \
+            const char *errMsg = hipGetErrorString(e);                                            \
+            std::ostringstream msg;                                                               \
+            msg << "Error " << e << "(" << errName << ") " << __FILE__ << ":" << __LINE__ << ": " \
+                << std::endl                                                                      \
+                << #expr << std::endl                                                             \
+                << errMsg << std::endl;                                                           \
+            throw std::runtime_error(msg.str());                                                  \
+        }                                                                                         \
+    } while (0)
+
+#define HIP_CHECK_RETURN(expr) \
+    do                         \
+    {                          \
+        hipError_t e = (expr); \
+        if (e)                 \
+            return e;          \
+    } while (0)
+
+struct KernelLaunchInfo
+{
+    std::string kernel, function, kernelHash;
+    int sharedMemoryBytes;
+    int gridX, gridY, gridZ;
+    int blockX, blockY, blockZ;
+};
+
+std::unordered_map<std::string, std::tuple<hipModule_t, hipFunction_t>> cache;
+using Int64Vector = std::vector<uint64_t>;
+using Int32Vector = std::vector<uint32_t>;
+
+int launch(const KernelLaunchInfo &info, const Int64Vector &tensors, const Int64Vector &dynamicDims)
+{
+    hipStream_t stream = at::hip::getCurrentHIPStream();
+    hipModule_t module;
+    hipFunction_t function;
```

Comment:
This "wave runtime" is using a generic name but is only supporting a single GPU API. IREE and iree-turbine both support CPU, Vulkan, CUDA, ROCm, Metal, and out of tree APIs. Extensions of the PyTorch ecosystem like iree-turbine could be limited to what PyTorch already supports (CPU, CUDA, ROCm, and a few other APIs IREE doesn't target), but we should at least be careful to set up an architecture that doesn't tie us directly to HIP/ROCm.

Reply:
Good point. I think this is an area where documentation can help where currently we only support MI, but could do more in the future.